### PR TITLE
split contract account

### PR DESCRIPTION
**Pull request by otselnik**

### CODE REVIEW

**Comment by otselnik** [comment]
test_event does not work at this moment
test_deploy, test_transaction and test_remove_caller are passing

**Comment by otselnik** [comment]
test_event, test_transaction and test_remove_caller need solana-cli from split_accoints_cli branch

File: evm_loader/program/src/account_data.rs [hunk 1]
```diff
@@ -11,23 +11,35 @@ pub struct AccountData {
     pub nonce: u8,
     pub trx_count: u64,
     pub signer: Pubkey,
+    pub code_account: Pubkey,
     pub code_size: u32,
 }
```
Comment (s-medvedev) [inline evm_loader/program/src/account_data.rs:22]: Create enum for hold information about AcountData, ContractData and Empty.

File: evm_loader/program/src/account_data.rs [hunk 2]
```diff
@@ -11,23 +11,35 @@ pub struct AccountData {
     pub nonce: u8,
     pub trx_count: u64,
     pub signer: Pubkey,
+    pub code_account: Pubkey,
     pub code_size: u32,
```
Comment (s-medvedev) [inline evm_loader/program/src/account_data.rs:20]: Move `code_size` field to ContractData

File: evm_loader/program/src/account_data.rs [hunk 3]
```diff
@@ -11,23 +11,35 @@ pub struct AccountData {
...
 }
 
 impl AccountData {
-    pub const SIZE: usize = 20+1+8+32+4;
+    pub const SIZE: usize = 20+1+8+32+32+4;
     pub fn size() -> usize {AccountData::SIZE}
+
+    pub fn get_code_account(src: &[u8]) -> Result<Pubkey, ProgramError> {
```
Comment (s-medvedev) [inline evm_loader/program/src/account_data.rs]: You can get this information from unpacked AccountData.

File: evm_loader/program/src/account_storage.rs [hunk 4]
```diff
@@ -19,22 +19,66 @@ pub struct ProgramAccountStorage<'a> {
...
         let mut accounts = Vec::with_capacity(account_infos.len());
         let mut aliases = Vec::with_capacity(account_infos.len());
+        
+        let mut contract_id: H160 = H160::zero();
+        let mut caller_id: H160 = H160::zero();
 
+        let mut skip_next: bool = false;
+        let mut contract: bool = true;
+        let mut caller: bool = true;
         for (i, account) in (&account_infos).iter().enumerate() {
```
Comment (s-medvedev) [inline evm_loader/program/src/account_storage.rs:39]: Rewrite with `while` cycle for account_infos.

File: evm_loader/program/src/entrypoint.rs [hunk 5]
```diff
@@ -130,10 +130,64 @@ fn process_instruction<'a>(
             debug_print!("create_account done");
             
             let mut data = program_info.data.borrow_mut();
-            let account_data = AccountData {ether, nonce, trx_count: 0u64, signer: *funding_info.key, code_size: 0u32};
+            let account_data = AccountData {ether, nonce, trx_count: 0u64, signer: *funding_info.key, code_account: Pubkey::new_from_array([0u8; 32]), code_size: 0u32};
             account_data.pack(&mut data)?;
             Ok(())
         },
+        EvmInstruction::CreateProgramAccount {lamports, space, ether, nonce} => {
```
Comment (s-medvedev) [inline evm_loader/program/src/entrypoint.rs]: Merge with CreateAccount instruction. You can check that user pass contract_info using next_account_info result.
Remove `space` argument.

File: evm_loader/program/src/entrypoint.rs [hunk 6]
```diff
@@ -130,10 +130,64 @@ fn process_instruction<'a>(
...
+
+            {// Do checks
+                let expected_address = Pubkey::create_program_address(&[ether.as_bytes(), &[nonce]], program_id)?;
+                if expected_address != *program_info.key {
+                    debug_print!("expected_address != *program_info.key");
+                    return Err(ProgramError::InvalidArgument);
+                };
+
+                let seed = bs58::encode(&ether.to_fixed_bytes()).into_string();
+                let expected_code_address = Pubkey::create_with_seed(&funding_info.key, &seed, &program_id)?;
```
Comment (s-medvedev) [inline evm_loader/program/src/entrypoint.rs]: This check not needed. But check that ContractInfo is empty now.

File: evm_loader/program/src/entrypoint.rs [hunk 7]
```diff
@@ -130,10 +130,64 @@ fn process_instruction<'a>(
...
+            let mut data = program_info.data.borrow_mut();
+            let account_data = AccountData {ether, nonce, trx_count: 0u64, signer: *funding_info.key, code_account: *program_code.key, code_size: 0u32};
+            account_data.pack(&mut data)?;
+
+            let mut code_data = program_code.data.borrow_mut();
+            code_data[..32].copy_from_slice(program_info.key.as_ref());
+
+            Ok(())
+        },
         EvmInstruction::CreateAccount2 {lamports, space, ether, nonce} => {
```
Comment (s-medvedev) [inline evm_loader/program/src/entrypoint.rs]: Remove this Instruction

File: evm_loader/program/src/entrypoint.rs [hunk 8]
```diff
@@ -185,15 +239,26 @@ fn process_instruction<'a>(
             if program_info.owner != program_id {
                 return Err(ProgramError::InvalidArgument);
             }
+
             do_write(program_info, offset, &bytes)
         },
+        EvmInstruction::WriteContractCode {offset, bytes} => {
```
Comment (s-medvedev) [inline evm_loader/program/src/entrypoint.rs]: Merge with `Write`. Add checks that user can change contract data.
1. empty account
2. contract without code (code_size = 0)

File: evm_loader/program/src/entrypoint.rs [hunk 9]
```diff
@@ -217,7 +282,7 @@ fn process_instruction<'a>(
             let mut account_storage = ProgramAccountStorage::new(program_id, accounts, accounts.last().unwrap())?;
```
Comment (s-medvedev) [inline evm_loader/program/src/entrypoint.rs]: Pass accounts without first entry.

File: evm_loader/program/src/solidity_account.rs [hunk 10]
```diff
@@ -38,48 +57,41 @@ impl<'a> SolidityAccount<'a> {
...
-        }*/
-        if self.account_data.code_size > 0 {
-            let data = self.data.borrow();
-            let offset = AccountData::SIZE;
-            let code_size = self.account_data.code_size as usize;
-            f(&data[offset..offset + code_size])
-        } else {
-            f(&[])
-        }
+    pub fn get_solana_address(&self) -> Pubkey {
```
Comment (s-medvedev) [inline evm_loader/program/src/solidity_account.rs:81]: Do not reorder functions

File: evm_loader/program/src/account_data.rs [hunk 11]
```diff
@@ -11,40 +11,112 @@ pub struct AccountData {
...
+    pub fn unpack(input: &[u8]) -> Result<Self, ProgramError> {
+        let (&tag, rest) = input.split_first().ok_or(ProgramError::InvalidAccountData)?;
+        Ok(match tag {
+            AccountType::EMPTY_TAG => AccountType::Empty,
+            AccountData::TAG => AccountType::AccountData( AccountData::unpack(rest)? ),
+            ContractData::TAG => AccountType::ContractData( ContractData::unpack(rest)? ),
+
+            _ => return Err(ProgramError::InvalidAccountData),
+        })
+    }
```
Comment (s-medvedev) [inline evm_loader/program/src/account_data.rs:42]: I would suggest adding `pack()`, `size()` and functions to this class.
`pack()` function should check the type of data saved in vector and raise an error if it changed from the current.

File: evm_loader/program/src/account_data.rs [hunk 12]
```diff
@@ -11,40 +11,112 @@ pub struct AccountData {
...
+}
+
+#[derive(Debug,Clone)]
+pub struct ContractData {
+    pub owner: Pubkey,
     pub code_size: u32,
 }
 
+#[derive(Debug,Clone)]
+pub enum AccountType {
```
Comment (s-medvedev) [inline evm_loader/program/src/account_data.rs]: I would suggest some renames:
AccountType -> AccountData
struct AccountData -> struct Account
struct ContractData -> struct Contract

File: evm_loader/program/src/instruction.rs [hunk 13]
```diff
@@ -256,15 +232,17 @@ impl<'a> EvmInstruction<'a> {
 pub fn on_return(
     myself_program_id: &Pubkey,
     status: u8,
-    mut result: Vec<u8>
+    result: &Vec<u8>
 ) -> Result<Instruction, ProgramError> {
-    result.insert(0, status);
-    result.insert(0, 6u8);
+    let mut data = Vec::new();
```
Comment (s-medvedev) [inline evm_loader/program/src/instruction.rs]: For better performance allocate space on vector initialization.

File: evm_loader/solana_utils.py [hunk 14]
```diff
@@ -205,7 +204,7 @@ def ether2program(self, ether):
         if isinstance(ether, str):
```
Comment (s-medvedev) [inline evm_loader/solana_utils.py:204]: You should get the implementation of `ether2programAddress` for `ether2program` function.
Also drop `ether2programAddress` function.

File: evm_loader/program/src/account_data.rs (hunk 11 above)
Comment (otselnik) [reply evm_loader/program/src/account_data.rs:42]: pushed work on this to
account_data_realisation_in_enum
branch

**Pull request closed without merging**